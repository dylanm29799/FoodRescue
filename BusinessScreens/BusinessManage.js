/*
 *
 * ClassName: BusinessManage.js
 *
 * Date: 28/04/2021
 *
 *
 * @author: Dylan Murphy, X17506166
 *
 * @reference : https://www.udemy.com/course/react-native-the-practical-guide/learn/lecture/15674818?start=0#overview
 * @reference : https://docs.expo.io/
 * @reference : https://firebase.google.com/docs/web/setup
 * @reference : https://github.com/wix/react-native-navigation
 * @reference : https://www.npmjs.com/package/react-native-dropdown-picker
 *
 *
 */
import React, { useEffect, useState } from "react";
import {
  View,
  Text,
  StyleSheet,
  FlatList,
  TouchableOpacity,
  Image,
} from "react-native";
import { scale } from "../components/ResponsiveText";
import DropDownPicker from "react-native-dropdown-picker";

import * as firebase from "firebase";

import { MaterialCommunityIcons } from "@expo/vector-icons";
const BusinessManage = (props) => {
  //Setting state
  const [orders, setOrders] = useState({});

  const dbconnection = firebase.firestore();
  //Setting state of querying
  const [showBy, setShowBy] = useState("All");

  console.log(showBy);
  var user = firebase.auth().currentUser;
  //Setting the initial colour
  var Colour = "#fff";
  useEffect(() => {
    //Getting all user orders
    console.log(user.uid);
    dbconnection
      .collection("OrderDetails")
      .where("businessID", "==", user.uid)
      .onSnapshot((querySnapshot) => {
        //Creating array
        var orders = [];
        //Pushing each order to the array
        querySnapshot.forEach(function (doc) {
          orders.push({
            ...doc.data(),
            key: doc.id,
          });
        });
        //Sorting array by time and date
        orders.sort(function (a, b) {
          return new Date(a.created.toDate()) - new Date(b.created.toDate());
        });
        //Pushing array to an object
        setOrders(orders);
      });
  }, []);
  //Rendering the order details
  const renderCategory = (itemData) => {
    var FinalDate;
    var Time;
    //Getting the time of the order
    var newCreated = itemData.item.created.toDate();
    var mins = newCreated.getMinutes();
    //Making minutes double digits if below 10 e.g 08 instead of 8
    if (newCreated.getMinutes() < 10) {
      mins = "0" + newCreated.getMinutes();
    }
    //Getting the date and time as user readable figures
    FinalDate = newCreated.getDate() + "/" + (newCreated.getMonth() + 1);
    Time = newCreated.getHours() + ":" + mins;
    //Getting image of product and first and last name of user
    //Setting the colour of the order
    if (itemData.item.Status == "In Progress") {
      Colour = "#ffc575";
    } else if (itemData.item.Status == "Completed") {
      Colour = "#d3f8d3";
    } else if (itemData.item.Status == "Cancelled") {
      Colour = "#ff6961";
    }
    //Displaying only the correct orders
    if (itemData.item.Status == showBy || showBy == "All") {
      return (
        <TouchableOpacity
          style={[styles.Categories, { backgroundColor: Colour }]}
          onPress={() => {
            props.navigation.navigate({
              //Navigating the the receipt page and passing the param of the order ID
              routeName: "businessReceipt",
              params: {
                orderID: itemData.item.key,
              },
            });
          }}
        >
          <View
            style={{
              flexDirection: "row",
              height: "100%",
              width: "100%",
              alignItems: "center",
              justifyContent: "center",
            }}
          >
            <View style={{ alignItems: "center" }}>
              <Text style={styles.text2}>
                {" "}
                {itemData.item.quantityOrdered}x {itemData.item.productName}
              </Text>
              <Text style={styles.text2}>{itemData.item.Status}</Text>
              <Text style={styles.text2}>
                {FinalDate.toString()} {Time}
              </Text>
              <Text style={styles.text2}>
                â‚¬{itemData.item.pricePerItem} each
              </Text>
            </View>
          </View>
        </TouchableOpacity>
      );
    }
  };

  return (
    <View style={{ flex: 1, backgroundColor: "white", alignItems: "center" }}>
      <View
        style={{
          flexDirection: "row",
          alignItems: "center",
          paddingVertical: 20,
          width: "100%",
          justifyContent: "center",
        }}
      >
        <Text style={{ fontSize: scale(12) }}>Sort By: </Text>
        <DropDownPicker
          //Dropdown picker allows the user to query the orders by status
          items={[
            {
              label: "All",
              value: "All",
              icon: () => (
                <MaterialCommunityIcons
                  name="all-inclusive"
                  size={24}
                  color="black"
                />
              ),
            },
            {
              label: "In Progress",
              value: "In Progress",
              icon: () => (
                <MaterialCommunityIcons
                  name="progress-alert"
                  size={24}
                  color="black"
                />
              ),
            },
            {
              label: "Completed",
              value: "Completed",
              icon: () => (
                <MaterialCommunityIcons
                  name="progress-check"
                  size={24}
                  color="black"
                />
              ),
            },
            {
              label: "Cancelled",
              value: "Cancelled",
              icon: () => (
                <MaterialCommunityIcons
                  name="progress-close"
                  size={24}
                  color="black"
                />
              ),
            },
          ]}
          defaultValue={"All"}
          containerStyle={{ height: 40, width: "50%" }}
          style={{ backgroundColor: "#fafafa" }}
          itemStyle={{
            justifyContent: "flex-start",
          }}
          dropDownStyle={{ backgroundColor: "#fafafa" }}
          onChangeItem={(item) => setShowBy(item.value)}
        />
      </View>
      <FlatList
        style={{ alignContent: "center", paddingTop: 20 }}
        data={orders}
        renderItem={renderCategory}
      />
    </View>
  );
};
//Stylesheet for styling
const styles = StyleSheet.create({
  Categories: {
    height: scale(120),
    alignItems: "center",
    justifyContent: "center",
    borderWidth: 1,
    width: scale(330),
    marginBottom: "3%",
    justifyContent: "space-evenly",
    borderRadius: 5,
  },
  text: {
    fontFamily: "MonL",
    fontSize: scale(16),
  },
  text2: {
    fontFamily: "MonM",
    fontSize: scale(18),
  },
});

export default BusinessManage;
