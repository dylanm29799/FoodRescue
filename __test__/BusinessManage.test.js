// __tests__/BusinessManage.test.js

import React from "react";
import BusinessManage from "../BusinessScreens/BusinessManage";
import renderer from "react-test-renderer";
import { firebaseConfig } from "../config";
import * as firebase from "firebase";

firebase.initializeApp(firebaseConfig);
test("renders correctly", () => {
  const tree = renderer.create(<BusinessManage />).toJSON();
  expect(tree).toMatchSnapshot();
});
