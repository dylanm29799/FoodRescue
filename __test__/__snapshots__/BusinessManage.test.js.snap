// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingVertical": 20,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 25.714285714285715,
        }
      }
    >
      Sort By: 
    </Text>
    <View
      style={
        Array [
          Object {
            "height": 40,
            "width": "50%",
          },
          Object {
            "zIndex": 5000,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onLayout={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fafafa",
            "borderBottomLeftRadius": 5,
            "borderBottomRightRadius": 5,
            "borderColor": "#dfdfdf",
            "borderTopLeftRadius": 5,
            "borderTopRightRadius": 5,
            "borderWidth": 1,
            "flex": 1,
            "flexDirection": "row",
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 5,
          }
        }
      >
        <Text />
        <Text
          style={
            Array [
              Object {
                "color": "red",
              },
              false,
              Object {
                "flex": 1,
                "opacity": 1,
              },
              Object {
                "marginLeft": 5,
                "marginRight": 0,
              },
              Object {},
              [Function],
              Object {},
            ]
          }
        >
          All
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderBottomLeftRadius": 0,
                "borderTopLeftRadius": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingVertical": 8,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {},
                Object {
                  "opacity": 1,
                },
              ]
            }
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderBottomLeftRadius": 5,
              "borderBottomRightRadius": 5,
              "borderColor": "#dfdfdf",
              "borderTopLeftRadius": 5,
              "borderTopRightRadius": 5,
              "borderWidth": 1,
              "paddingHorizontal": 10,
              "paddingVertical": 5,
            },
            Object {
              "alignItems": "center",
              "borderTopLeftRadius": 0,
              "borderTopRightRadius": 0,
              "justifyContent": "center",
              "position": "absolute",
              "width": "100%",
            },
            Object {
              "backgroundColor": "#fafafa",
            },
            Object {
              "borderWidth": 0,
              "display": "none",
              "position": "relative",
            },
            Object {
              "maxHeight": 150,
              "top": undefined,
              "zIndex": 5000,
            },
          ]
        }
      >
        <RCTScrollView
          nestedScrollEnabled={true}
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View>
            <View
              onLayout={[Function]}
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingVertical": 8,
                    "width": "100%",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignContent": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                    }
                  }
                >
                  <Text />
                  <Text
                    style={
                      Array [
                        Object {},
                        Object {},
                        Object {
                          "marginHorizontal": 5,
                        },
                      ]
                    }
                  >
                    All
                  </Text>
                </View>
              </View>
            </View>
            <View
              onLayout={[Function]}
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingVertical": 8,
                    "width": "100%",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignContent": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                    }
                  }
                >
                  <Text />
                  <Text
                    style={
                      Array [
                        Object {},
                        false,
                        Object {
                          "marginHorizontal": 5,
                        },
                      ]
                    }
                  >
                    In Progress
                  </Text>
                </View>
              </View>
            </View>
            <View
              onLayout={[Function]}
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingVertical": 8,
                    "width": "100%",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignContent": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                    }
                  }
                >
                  <Text />
                  <Text
                    style={
                      Array [
                        Object {},
                        false,
                        Object {
                          "marginHorizontal": 5,
                        },
                      ]
                    }
                  >
                    Completed
                  </Text>
                </View>
              </View>
            </View>
            <View
              onLayout={[Function]}
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingVertical": 8,
                    "width": "100%",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignContent": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                    }
                  }
                >
                  <Text />
                  <Text
                    style={
                      Array [
                        Object {},
                        false,
                        Object {
                          "marginHorizontal": 5,
                        },
                      ]
                    }
                  >
                    Cancelled
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={Object {}}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "alignContent": "center",
        "paddingTop": 20,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
